### DISCUSSION TITLE
ESM issue: rel not being exported

### DISCUSSION THREAD

**Discussion by bakura10** [mikro-orm/mikro-orm#4915] [General]
Hi,

I'm trying to do another try at migrating to ESM, but I'm still pulling my hair on an issue (version tested 6.0.0-rc.0).

Here is a reproduction case:

src/entities/one.ts

```
import { Entity, PrimaryKey, OneToOne, Rel } from "@mikro-orm/postgresql";
import { Two } from "./two.js";

@Entity()
export class One {
  @PrimaryKey()
  id: string = crypto.randomUUID();

  @OneToOne()
  two?: Rel<Two>;
}
```

src/entities/two.ts:

```
import { Entity, PrimaryKey } from "@mikro-orm/postgresql";

@Entity()
export class Two {
  @PrimaryKey()
  id: string = crypto.randomUUID();
}
```

src/mikro-orm.config.ts:

```
import { defineConfig } from "@mikro-orm/postgresql";
import { One } from "./entities/one.js";
import { Two } from "./entities/two.js";

export default defineConfig({
  host: process.env.DB_HOST,
  port: 5432,
  dbName: process.env.DB_NAME,
  user: process.env.DB_USERNAME,
  password: process.env.DB_PASSWORD,
  entities: [One, Two],
  debug: true
});
```

ts-config.json:

```
{
  "compilerOptions": {
    "target": "ES2020",
    "moduleResolution": "NodeNext",
    "module": "NodeNext",
    "outDir": "./dist",
    "allowJs": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "strict": true,
    "skipLibCheck": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false
  },
  "ts-node": {
    "esm": true,
    "transpileOnly": true
  },
  "include": [
    "src/**/*.ts",
  ],
}
```

package.json:

```
{
  "type": "module",
  "scripts": {
    "orm": "mikro-orm-esm"
  },
  "devDependencies": {
    "@mikro-orm/cli": "6.0.0-rc.0",
    "ts-node": "10.9.1"
  },
  "dependencies": {
    "@mikro-orm/core": "6.0.0-rc.0",
    "@mikro-orm/postgresql": "6.0.0-rc.0"
  },
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts"
    ]
  }
}
```

Without using the `Rel`, I face the issue outlined in doc:

```
configuration not found (Cannot access 'Two' before initialization)
```

After using Rel, I face this issue:

```
configuration not found (The requested module '@mikro-orm/postgresql' does not provide an export named 'Rel')
```

I tried to change to use @mikro-orm/core instead, but I face the same issue.

I'm using the `mikro-orm-esm`, but using the `mikro-orm` binary adds even more issue.

Would someone have an idea on the root cause here?

**Comment by B4nan** [comment]
It's a type, so a type import should help here. I am really curious why this happens, I never came across this problem myself but got several reports during the last year or so.

```ts
import { Entity, PrimaryKey, OneToOne, type Rel } from "@mikro-orm/postgresql";
```

**Reply by bakura10** [reply]
Thanks for the answer!

Using type indeed helped remove the error, but in turns it brought back the:

```
configuration not found (Cannot access 'Two' before initialization)
```

Those ESM config will drive me crazy...

**Reply by B4nan** [reply]
I dont think it can help, but you surely need to provide the type explicitly too:

```ts
@OneToOne({ entity: () => Two, nullable: true })
two?: Rel<Two>;
```

Or do you use ts-morph? If so, you can disable the `emitDecoratorMetadata` completely.

Anyway, as mentioned, I never saw this happening in my projects, do you use anything special like esbuild, swc, babel? The [new guide](https://github.com/mikro-orm/guide) uses ESM with ts-morph and doesn't suffer from this, just tried to enable `emitDecoratorMetadata` but things still work as expected. 

Complete repro welcome.

**Reply by bakura10** [reply]
I manage to make it work. It was indeed the entity missing in the associations. I will send a doc PR next week to better clarify this :)

**Reply by B4nan** [reply]
Ok so I actually am able to reproduce this, but only when running the code via ts-node with `emitDecoratorMetadata` enabled.

I can get things working if I remove the `transpileOnly` option and ensure the problematic type is not used anywhere without the `Rel` wrapper (I had it also in the constructor parameters which caused the problems, it wasn't about the property definition at all in my case).

**Reply by bakura10** [reply]
Should the Rel also be used for Embedded or only for OneToOne/ManyToOne/OneToMany associations?

**Reply by B4nan** [reply]
its identity type, it doesnt really matter where you use it, all it does is to get around cycles with reflect metadata, and you probably wont fall into those with embedded properties
